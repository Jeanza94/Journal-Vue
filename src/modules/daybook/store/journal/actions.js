import journalApi from "@/api/journalApi"


export const loadEntries = async ( { commit } ) => {
    try {
        const { data } = await journalApi.get('/entries.json');
        const idArr = Object.keys(data);
        const entries = idArr.map(id => {
            return {
                id,
                ...data[id]
            }
        });
        
        commit('setEntries', entries);
    } catch (error) {
        commit('setEntries', [])
    }
    
}

export const updateEntry = async ({ commit }, entry = {}) => {
    const { id, ...rest } = entry;
    const { data } = await journalApi.put(`/entries/${ id }.json`, {...rest})
    
    if (!data) return;

    data.id = id;
    commit('updateEntry', data);
    return data;

}

export const createEntry = async ({ commit }, entry = {}) => {
    if ( entry.text.trim().length === 0 ) {
        console.log(entry.text)
        return console.log('Escribe algo en el texto')
    }
    const { data } = await journalApi.post(`/entries.json`, entry );

    if (!data) {
        return console.log('No se pudo crear en firebase')
    }

    entry.id = data.name;

    commit('addEntry', entry)
    return entry;
}

export const deleteEntry = async({ commit }, id = '') => {
    try {
        await journalApi.delete(`/entries/${id}.json`)
        commit('deleteEntry', id)
    } catch (error) {   
        console.log(error);
    }

}