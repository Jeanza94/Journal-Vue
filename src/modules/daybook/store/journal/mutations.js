
export const setEntries = ( state, entries ) => {
    state.entries = [...state.entries, ...entries];
    state.isLoading = false;
}

export const updateEntry = ( state, entryUpdated ) => {
    
    state.entries = state.entries.map(entry => {
        if ( entry.id === entryUpdated.id ) {
            return entryUpdated
        }
        return entry;
    });
    
    
}

export const addEntry = ( state, entry ) => {
    state.entries = [entry, ...state.entries];
}

export const deleteEntry = (state, id) => {
    state.entries = state.entries.filter(entry => entry.id !== id)
}