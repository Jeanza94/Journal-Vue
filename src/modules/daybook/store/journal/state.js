

export default () => {
    return {
        isLoading: true,
        entries: [],
    }
}