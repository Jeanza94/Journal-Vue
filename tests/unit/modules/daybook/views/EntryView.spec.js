import { shallowMount } from "@vue/test-utils";
import { createStore } from "vuex";
import EntryView from '@/modules/daybook/views/EntryView'
import journal from "@/modules/daybook/store/journal";
import { journalState } from "../../../mock-data/test-journal-state";
import Swal from 'sweetalert2'

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    close: jest.fn(),
}))

const createVuexStore = ( initialState ) => {
    return createStore({
        modules:{
            journal: {
                ...journal,
                state: {...initialState}
            }
        }
    })
}

describe('Pruebas en el componente EntryView', () => { 

    const store = createVuexStore(journalState)
    store.dispatch = jest.fn();
    const mockRouter = {
        push: jest.fn()
    }

    let wrapper;

    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallowMount(EntryView, {

            props: {
                id: journalState.entries[0].id
            },

            global: {
                mocks: {
                    $router: mockRouter
                },

                plugins: [store]
            }
        })
    })

    test('debe de sacar al usuario porque el id no existe', () => { 

        const wrapper = shallowMount(EntryView, {
            
            props: {
                id:'Este id no es valido'
            },

            global: {
                mocks: {
                    $router: mockRouter
                },

                plugins: [ store ]
            }
        } )

        expect(mockRouter.push).toHaveBeenCalled();
        expect(mockRouter.push).toHaveBeenCalledWith({ name: 'no-entry' })
        
        
    });
    
    test('debe de mostrar la entrada correctamente', () => { 
        expect(wrapper.html()).toMatchSnapshot();
        expect(mockRouter.push).not.toHaveBeenCalled();
        
    });

    test('debe de borrar la entrada y salida', (done) => { 
        jest.setTimeout(10000)
        Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: true}))
        wrapper.find('.btn-danger').trigger('click');

        expect(Swal.fire).toHaveBeenCalledWith({
            title: 'Â¿Esta seguro?',
            text: 'Una vez borrado, no se puede recuperar',
            showDenyButton: true,
            confirmButtonText: 'Si estoy seguro'
        })

        setTimeout(() => {
            expect(store.dispatch).toHaveBeenCalled();
            expect(mockRouter.push).toHaveBeenCalledWith({name: 'no-entry'})
            done()
        }, 1);
        


    });

});