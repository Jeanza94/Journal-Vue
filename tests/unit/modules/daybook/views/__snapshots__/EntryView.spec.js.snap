// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente EntryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">6</span><span class="mx-1 fs-3">Diciembre</span><span class="mx-2 fs-4 fw-light">2022, Martes</span></div><input type="file" accept="image/png, image/jpeg" style="display: none;">
  <div><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="Â¿Que sucedio hoy?"></textarea></div>
<!--v-if-->
<!--v-if-->
<fab-button-stub icon="fa-save"></fab-button-stub>
`;
